working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,485 @@
+type parameters _T_:
+
+1. _A_ is a function literal expression,
+
+2. AND the parameter in the invocation target corresponding to _A_ is function
+   typed,
+
+3. AND _T_ is a free variable in the type of at least one of the parameters of
+   that function type,
+
+4. AND the corresponding parameter in _A_ does not have a type annotation,
+
+5. AND EITHER:
+
+   * The parameter in the invocation target corresponding to _B_ is function
+     typed, and _T_ is a free variable in its return type
+
+   * OR the parameter in the invocation target corresponding to _B_ is _not_
+     function typed, and _T_ is a free variable in its type.
+
+_The idea here is that we're trying to draw a dependency edge from A to B in
+precisely those cirumstances in which there is likely to be a benefit to
+performing a round of horizontal inference after type inferring A and before
+type inferring B._
+
+If the type of the invocation target is `dynamic` or the `Function` class, or
+some non-generic function type, then the resulting graph has no edges.
+
+_So, for example, if the invocation in question is this:_
+
+```dart
+f((t, u) { ... } /* A */, () { ... } /* B */, (v) { ... } /* C */, (u) { ... } /* D */)
+```
+
+_And the target of the invocation is declared like this:_
+
+```dart
+void f<T, U, V>(
+    void Function(T, U) a,
+    T b,
+    U Function(V) c,
+    V Function(U) d) => ...;
+```
+
+_then the resulting dependency graph looks like this:_
+
+&emsp;B &lArr; A &rArr; C &hArr; D
+
+_(That is, there are four edges, one from A to B, one from A to C, one from C to D, and one from D to C)._
+
+#### Stage selection
+
+After building the dependency graph, we condense it into its strongly connected
+components (a.k.a. "dependency cycles").  The resulting condensation is
+[guaranteed to be acyclic] (that is, considering the nodes of the condensed
+graph, the arguments in each node depend transitively on each other, and
+possibly on the arguments in other nodes, but no dependency cycle exists between
+one node and another).
+
+[guaranteed to be acyclic]: https://en.wikipedia.org/wiki/Strongly_connected_component#Definitions
+
+_For the example above, C and D are condensed into a single node, so the graph
+now looks like this:_
+
+&emsp;{B} &lArr; {A} &rArr; {C, D}
+
+Now, the nodes are grouped into stages as follows.  Stage zero consists of the
+arguments from all nodes that have no incoming edges.  Then, those nodes are
+removed from the graph, along with all their outgoing edges.  For each of the
+following stages, we follow the same procedure: from the graph produced by stage
+_k_, let stage _k+1_ be the set of arguments from all nodes that have no
+incoming edges, then delete those nodes and their outcoming edges to produce a
+newly reduced graph.  We repeat this until the graph is empty.
+
+_In this example, that means that there will be two stages.  The first stage will
+consist of argument A, and the second stage will consist of arguments B, C, and
+D._
```

Comment:
I'd expect the opposite order.

As described (around lines 290-306), two nodes (each node is an actual argument) A and B have a dependency if certain criteria are satisfied, and as described the edge goes from A to B. However, the information flow goes from B to A, because it is B that may contain information after inference which is useful during inference on A. So we want to analyze B before A.

But with the graph {B} ⇐ {A} ⇒ {C, D}, it is stated (line 362) that the first stage consists of {A} and the second one consists of {B,C,D}.

I would expect the first (zeroth) stage to be {B,C,D} and the next one {A}.

Reply:
You're right.  Fixed.